apps/spruce/src/pages/task/taskTabs/TaskHistory/caching/index.ts

```diff
@@ -0,0 +1,170 @@
+
+  // Always query for date, since it's an uncommon operation.
+  const date = args?.options?.date ?? "";
+  if (date) {
+    return undefined;
+  }
+
+  const cursorId = args?.options?.cursorParams?.cursorId ?? "";
+  const includeCursor = args?.options?.cursorParams?.includeCursor ?? false;
+  const direction = args?.options?.cursorParams?.direction ?? "";
+  const limit = args?.options?.limit ?? ACTIVATED_TASKS_LIMIT;
+
+  const pagination = readField<TaskHistoryPagination>("pagination", existing);
+  const { mostRecentTaskOrder = 0, oldestTaskOrder = 0 } = pagination ?? {};
+
+  const existingTasks = readField<TaskHistoryTask[]>("tasks", existing) ?? [];
+
+  const idx = existingTasks.findIndex((t) => {
+    const taskId = readField<string>("id", t) ?? "";
+    return taskId === cursorId;
+  });
+
+  if (idx === -1) {
+    return undefined;
+  }
+
+  const pagingBackwards = direction === TaskHistoryDirection.After;
+  const pagingForwards = direction === TaskHistoryDirection.Before;
+  const activeTasks = new Set();
+
+  let startIndex = 0;
+  let endIndex = 0;
+
+  if (pagingBackwards) {
+    startIndex = 0;
+    endIndex = includeCursor ? idx : idx - 1;
+
+    // Count backwards for paginating backwards.
+    for (let i = endIndex; i >= 0; i--) {
+      const activated =
+        readField<boolean>("activated", existingTasks[i]) ?? false;
+      if (activated) {
+        const id = readField<string>("id", existingTasks[i]) ?? "";
+        activeTasks.add(id);
+      }
+
+      // When we've collected LIMIT active tasks, we're done.
+      // We should also grab any leading inactive tasks.
+      if (activeTasks.size === limit) {
+        startIndex = i;
+        if (isAllInactive(existingTasks.slice(0, startIndex), readField)) {
+          startIndex = 0;
+        }
+        break;
+      }
+    }
+  }
+
+  if (pagingForwards) {
+    startIndex = includeCursor ? idx : idx + 1;
+    endIndex = existingTasks.length;
+
+    // Count forwards for paginating forwards.
+    for (let i = startIndex; i < existingTasks.length; i++) {
+      const activated =
+        readField<boolean>("activated", existingTasks[i]) ?? false;
+      if (activated) {
+        const id = readField<string>("id", existingTasks[i]) ?? "";
+        activeTasks.add(id);
+      }
+
+      // When we've collected LIMIT active tasks, we're done.
+      // We should also grab any trailing inactive tasks.
+      if (activeTasks.size === limit) {
+        endIndex = i;
+        if (isAllInactive(existingTasks.slice(endIndex + 1), readField)) {
+          endIndex = existingTasks.length;
+        }
+        break;
+      }
+    }
+  }
+
+  if (activeTasks.size < limit) {
+    const allTaskOrders =
+      readField<Set<number>>("allTaskOrders", existing) ?? new Set<number>();
+    if (pagingBackwards && !allTaskOrders.has(mostRecentTaskOrder)) {
+      return undefined;
+    }
+    if (pagingForwards && !allTaskOrders.has(oldestTaskOrder)) {
+      return undefined;
+    }
+  }
```

Comment:
(optional comment) This has a bit of repition and I think we can extract this out into utils and DRY out the logic a bit.

```suggestion
  let endIndex = existingTasks.length - 1;
  let pagingDirection: Direction | null = null;

  if (direction === TaskHistoryDirection.After) pagingDirection = "backwards";
  if (direction === TaskHistoryDirection.Before) pagingDirection = "forwards";

  if (pagingDirection) {
    ({ endIndex, startIndex } = getPaginationIndices(
      [...existingTasks],
      readField,
      idx,
      limit,
      pagingDirection,
      includeCursor,
    ));
  }

  // Check if we have enough active tasks
  const activeTasksCount = existingTasks
    .slice(startIndex, endIndex + 1)
    .filter((t) => readField<boolean>("activated", t)).length;

  if (activeTasksCount < limit) {
    const allTaskOrders =
      readField<Set<number>>("allTaskOrders", existing) ?? new Set<number>();
    if (
      (pagingDirection === "backwards" &&
        !allTaskOrders.has(mostRecentTaskOrder)) ||
      (pagingDirection === "forwards" && !allTaskOrders.has(oldestTaskOrder))
    ) {
      return undefined;
    }
  }
```

Reply:
These are the utility functions for the suggestion above.
```ts
type Direction = "backwards" | "forwards";
interface Indices {
  startIndex: number;
  endIndex: number;
}

const getPaginationIndices = (
  tasks: TaskHistoryTask[],
  readField: ReadFieldFunction,
  cursorIdx: number,
  limit: number,
  direction: Direction,
  includeCursor: boolean,
): Indices =>
  direction === "backwards"
    ? paginateBackwards(tasks, readField, cursorIdx, limit, includeCursor)
    : paginateForwards(tasks, readField, cursorIdx, limit, includeCursor);

const paginateBackwards = (
  tasks: TaskHistoryTask[],
  readField: ReadFieldFunction,
  cursorIdx: number,
  limit: number,
  includeCursor: boolean,
): Indices => {
  const startAt = includeCursor ? cursorIdx : cursorIdx - 1;
  const boundary = findBoundary(tasks, readField, startAt, -1, limit);
  const endIndex = startAt;
  const startIndex =
    boundary == null || isAllInactive(tasks.slice(0, boundary), readField)
      ? 0
      : boundary;
  return { startIndex, endIndex };
};

const paginateForwards = (
  tasks: TaskHistoryTask[],
  readField: ReadFieldFunction,
  cursorIdx: number,
  limit: number,
  includeCursor: boolean,
): Indices => {
  const startAt = includeCursor ? cursorIdx : cursorIdx + 1;
  const boundary = findBoundary(tasks, readField, startAt, +1, limit);
  const startIndex = startAt;
  const endIndex =
    boundary == null || isAllInactive(tasks.slice(boundary + 1), readField)
      ? tasks.length - 1
      : boundary;
  return { startIndex, endIndex };
};

const findBoundary = (
  tasks: TaskHistoryTask[],
  readField: ReadFieldFunction,
  idx: number,
  step: 1 | -1,
  limit: number,
): number | null => {
  const seenIds = new Set<string>();
  for (let i = idx; i >= 0 && i < tasks.length; i += step) {
    if (readField<boolean>("activated", tasks[i])) {
      seenIds.add(readField<string>("id", tasks[i]) ?? "");
      if (seenIds.size === limit) return i;
    }
  }
  return null;
};

```